Make Whisper Encoder's sinusoidal PE non-trainable by default

gau-nernst opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes #25989 

I'm not too familiar with Jax/Flax and can't find a simple way to set a variable a non-trainable in Flax. Do advise on how I should approach this.

Should we have a test for this behavior also? i.e. test that Whisper Encoder PE is non-trainable by default.

Another note. Should Encoder's positional encodings be initialized with sinusoids? Just like the official repo

https://github.com/openai/whisper/blob/main/whisper/model.py#L150

```python
def sinusoids(length, channels, max_timescale=10000):
    """Returns sinusoids for positional embedding"""
    assert channels % 2 == 0
    log_timescale_increment = np.log(max_timescale) / (channels // 2 - 1)
    inv_timescales = torch.exp(-log_timescale_increment * torch.arange(channels // 2))
    scaled_time = torch.arange(length)[:, np.newaxis] * inv_timescales[np.newaxis, :]
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerz and @pacman100

Integrations:

- deepspeed: HF Trainer/Accelerate: @pacman100
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc and @younesbelkada

Documentation: @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

@sanchit-gandhi

gau-nernst committed:
set encoder's PE as non-trainable

Commit:
freeze flax

sanchit-gandhi commented:
Hey @gau-nernst - thanks very much for opening this PR! Looks like a great start already. I pushed the Flax changes in the latest commit. In short, the simplest way of setting the parameters to un-trainable in Flax is by stopping the back-prop through the layers. Otherwise, we need to explicitly pass a dict to the optimiser that defines which parameters are trainable/non-trainable (see https://colab.research.google.com/drive/1K-5bz6R6kt9GAvaUHvzYvvA-IOAO2PhL#scrollTo=BrF6Dtb8GlkJ)

There's not a test to check that the embed params are non-trainable, but you could certainly add one. This could follow the style of test that we use to check that we correctly freeze the encoder when we do decoder-only fine-tuning:
https://github.com/huggingface/transformers/blob/ce2e7ef3d96afaf592faf3337b7dd997c7ad4928/tests/models/whisper/test_modeling_whisper.py#L338

Regarding initialising the weights with sinusoidal embeddings - I agree that this should be the default case! In 99% of cases users will just use the model from pre-trained, in which case the embeddings will be initialised with the sinusoids, but if a user were to randomly initialise the model, the embeddings would be initialised incorrectly.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_26032). All of your documentation changes will be reflected on that endpoint.

gau-nernst commented:
That's a great solution! However, from what I understand, it means that in the Flax implementation, it is not possible (or easy) to re-enable training for positional encodings? (something that we discussed previously)

sanchit-gandhi commented:
It's possible (but a bit involved) to add functionality to toggle whether we train the PE's in Flax. However, to me this PR is a bug fix, rather than a feature addition. I agree with what you said in the issue that we should not train the embeddings, since they used fixed sinusoidal embeddings in the original implementation, so I think it's fine if we do a straight fix and always freeze the embeddings here, since this is the correct behaviour.

gau-nernst committed:
init sinusoids

gau-nernst committed:
add test for non-trainable embed positions

gau-nernst committed:
simplify TF encoder embed_pos

gau-nernst committed:
Merge branch 'main' into whisper_encoder_pe

gau-nernst committed:
revert tf

gau-nernst commented:
I added sinusoids weight init for PyTorch implementation. Looking at TF and Flax, I'm not sure where to put weight init. It seems like there is no weight init code in TF? For Flax, I see this but don't really understand what's going on.

https://github.com/huggingface/transformers/blob/0a55d9f7376f72ad3ff296d4249840021b03bcc4/src/transformers/models/whisper/modeling_flax_whisper.py#L865-L895

From other TF Keras and Flax code I have seen, I think the typical pattern is to pass weight init function to a module when it is created? I'm not sure what is the pattern HF is using here.

gau-nernst committed:
clean up

sanchit-gandhi commented:
The `init_weights` function in Flax is used to initialise the Flax model's parameters by passing a set of dummy inputs (zeros and ones). Flax traces out the shapes of the weights that you get when you pass these dummy inputs, and initialises weights with the right shapes accordingly (see https://flax.readthedocs.io/en/latest/guides/flax_basics.html#model-parameters-initialization).

This `init_weights` function won't actually change the values of the weights, just their shapes and dtypes. To change the initialising function, we can pass an argument `embedding_init` function to the init of the embedding layer: https://flax.readthedocs.io/en/latest/api_reference/flax.linen/_autosummary/flax.linen.Embed.html#flax.linen.Embed.embedding_init

The init function should be an instance of a JAX initialiser. That is, it should take the PRNG Key as the first argument, as well as the shape and target dtype of the module: https://jax.readthedocs.io/en/latest/jax.nn.initializers.html

gau-nernst committed:
Merge branch 'main' into whisper_encoder_pe

gau-nernst committed:
add sinusoidal init for jax

gau-nernst committed:
make consistent sinusoidal function

gau-nernst committed:
fix dtype

gau-nernst committed:
add default dtype

gau-nernst committed:
use numpy for sinusoids. fix jax

gau-nernst committed:
add sinusoid init for TF

gau-nernst committed:
fix

gau-nernst committed:
use custom embedding

sanchit-gandhi reviewed (commented):
Thanks for your follow-up work on this issue @gau-nernst! Nice job on getting the Flax and TF parts working as well 👏 I've left some suggestions below on how we could potentially re-factor the code a bit to make it as clear as possible for the final PR, let me know if you have any questions!

gau-nernst commented on src/transformers/models/whisper/modeling_flax_whisper.py:
```diff
@@ -58,6 +60,16 @@
 
 
+# Copied from transformers.models.whisper.modeling_whisper.sinusoids
+def sinusoids(length: int, channels: int, max_timescale: float = 10000) -> np.ndarray:
```
Sure, that works as well. I used numpy initially to generate the sinusoids so that we can copy it across the 3 files and avoid errors. But having separate functions is fine by me too.

gau-nernst commented on src/transformers/models/whisper/modeling_whisper.py:
```diff
@@ -665,7 +674,12 @@ def _init_weights(self, module):
-            module.weight.data.normal_(mean=0.0, std=std)
+            if not module.weight.requires_grad:
+                # sinusoidal positional encodings used in WhisperEncoder
+                with torch.no_grad():
```
I'm not 100% sure when `_init_weights()` is called. If it is always called at the end of `__init__()` (in `post_init()`?), and we don't freeze anything else in `__init__()`, it should still work as intended. However, I agree that relying on this behavior is error-prone and not exactly clean.

I don't think there is a clean way for `_init_weights()` to know an `nn.Embedding()` layer is from the Encoder? If that is the case, I think the best solution is to initialize sinusoids after `_init_weights()` is called within `__init__()`?

gau-nernst committed:
use specialized init for each impl

gau-nernst committed:
fix sinusoids init. add test for pytorch

gau-nernst committed:
fix TF dtype

sanchit-gandhi commented on src/transformers/models/whisper/modeling_flax_whisper.py:
```diff
@@ -58,6 +59,22 @@
 remat = nn_partitioning.remat
 
 
+def sinusoidal_embedding_init(max_timescale: float = 10000):
```
Sorry @gau-nernst, can we not just define one function here? We should move away from defining two functions, where the outer one just calls the inner one directly

gau-nernst replied:
The reason I do it like this is to allow changing the default `max_timescale` value ("parameterized" function), so that it is on feature-parity with the PyTorch version. If we just define 1 function, there is no way to change `max_timescale` value, since Jax/Keras will only call the function with (shape, dtype) (and additionally RNG for Jax) (technically we can still bypass this by using `functools.partial()`). I follow Jax for this design (https://github.com/google/jax/blob/3247db774ea387098bd9d9049886030dc666cb39/jax/_src/nn/initializers.py#L133-L157). Another way is to make it a class (like Keras does).

Realistically the users won't be able to specify `max_timescale` to the model anyway since we don't expose it. So it would be fine for me to make `max_timescale` as a hard-coded constant also.

sanchit-gandhi replied:
Yeah if `max_timescales` is not a reachable argument by the user let's just hardcode it. We tend to do this anyway for sinusoidal embeddings: https://github.com/huggingface/transformers/blob/391177441b133645c02181b57370ab12f71b88c4/src/transformers/models/gptj/modeling_flax_gptj.py#L109-L110

sanchit-gandhi commented on src/transformers/models/whisper/modeling_whisper.py:
```diff
@@ -835,13 +847,16 @@ def __init__(self, config: WhisperConfig):
         # Initialize weights and apply final processing
         self.post_init()
+        with torch.no_grad():
+            self.embed_positions.weight.copy_(sinusoids(self.max_source_positions, embed_dim))
```
This was more appropriate in `_init_weights`! Let's keep it there

sanchit-gandhi replied:
This still holds - can this go in `_init_weights` if possible?

sanchit-gandhi commented on tests/models/whisper/test_modeling_whisper.py:
```diff
@@ -351,6 +351,13 @@ def test_requires_grad_with_frozen_encoder(self):
             self.assertFalse(all(encoder_grads))
             self.assertTrue(all(decoder_grads))
 
+    def test_requires_grad_encoder_embed_positions(self):
```
Could we still add an integration test of some sort? One where we check that the embeddings for the encoder actually match the expected sinusoid values?

sanchit-gandhi reviewed (commented):
Thanks for updating the `sinusoids` functions to their respective libraries! They look a lot better. Just a few small comments regarding the initialisation :)

gau-nernst commented on tests/models/whisper/test_modeling_whisper.py:
```diff
@@ -351,6 +351,20 @@ def test_requires_grad_with_frozen_encoder(self):
+            encoder = model.get_encoder()
+            self.assertFalse(encoder.embed_positions.weight.requires_grad)
+
+    def test_encoder_sinusoidal_embed_positions(self):
```
@sanchit-gandhi I have already added the test for sinusoidal embeddings here. Not done yet for Jax/TF though.

sanchit-gandhi replied:
Cool! Looks good!

gau-nernst committed:
simplify sinusoid init for flax and tf

gau-nernst committed:
add tests for TF

gau-nernst committed:
change default dtype to float32

gau-nernst committed:
add sinusoid test for flax

gau-nernst commented:
@sanchit-gandhi I fixed the embedding init for TF and Flax as you requested. I also add test for TF and Flax. For Flax, I don't add a test for non-trainable sinusoidal embedding, because I don't know how to do it cleanly. For checking the weight init in Flax, I don't know Flax semantics so well, so I added a rather "crude" solution to get the encoder position embeddings.

sanchit-gandhi commented on src/transformers/models/whisper/modeling_flax_whisper.py:
```diff
@@ -58,6 +59,20 @@
 remat = nn_partitioning.remat
 
 
+def sinusoidal_embedding_init(key, shape, dtype=jnp.float_) -> jax.Array:
```
Really nice!

sanchit-gandhi commented on src/transformers/models/whisper/modeling_flax_whisper.py:
```diff
@@ -58,6 +59,20 @@
+        raise ValueError(
+            f"Number of channels has to be divisible by 2 for sinusoidal positional embeddings, got {channels} channels."
+        )
+    max_timescale = 10000
+    log_timescale_increment = math.log(max_timescale) / (channels // 2 - 1)
```
(nit) maybe cleaner to define it on one line
```suggestion
    log_timescale_increment = math.log(10000) / (channels // 2 - 1)
```

sanchit-gandhi commented on src/transformers/models/whisper/modeling_flax_whisper.py:
```diff
@@ -673,6 +694,8 @@ def __call__(
         hidden_states = jax.nn.gelu(self.conv2(hidden_states), approximate=False)
 
         embed_positions = self.embed_positions(jnp.arange(self.config.max_source_positions))
+        # freeze the sinusoidal embeddings by stopping the back-prop
```
Note to reviewer: by default we freeze the embeddings in Flax, and don't provide an override. See this explanation for detail: https://github.com/huggingface/transformers/pull/26032#issuecomment-1719537609

sanchit-gandhi commented on src/transformers/models/whisper/modeling_tf_whisper.py:
```diff
@@ -59,6 +59,20 @@
+        raise ValueError(
+            f"Number of channels has to be divisible by 2 for sinusoidal positional embeddings, got {channels} channels."
+        )
+    max_timescale = 10000
+    log_timescale_increment = math.log(max_timescale) / (channels // 2 - 1)
```
```suggestion
    log_timescale_increment = math.log(10000) / (channels // 2 - 1)
```

sanchit-gandhi commented on tests/models/whisper/test_modeling_flax_whisper.py:
```diff
@@ -387,6 +388,19 @@ def test_save_load_to_base(self):
                     max_diff = (base_params[key] - base_params_from_head[key]).sum().item()
                     self.assertLessEqual(max_diff, 1e-3, msg=f"{key} not identical")
 
+    def test_encoder_sinusoidal_embed_positions(self):
```
To test that the Flax embeddings are non-trainable (frozen), you can follow this Flax Wav2Vec2 test: https://github.com/huggingface/transformers/blob/4b4c6aabfb054fd42e5737ffafc56739adf3cce4/tests/models/wav2vec2/test_modeling_flax_wav2vec2.py#L281

To me this is optional: we know that the embeddings are initialised correctly through your test, and that grads are set to zero by action of `jax.lax.stop_gradient`, so up to you if you want to add this!

sanchit-gandhi reviewed (commented):
Very nice @gau-nernst - especially the Flax init which is really clean now 👌 Could the PT init go in `_init_weights`? Otherwise it all looks good to me!

gau-nernst committed:
Update src/transformers/models/whisper/modeling_flax_whisper.py

Co-authored-by: Sanchit Gandhi <93869735+sanchit-gandhi@users.noreply.github.com>

gau-nernst committed:
Update src/transformers/models/whisper/modeling_tf_whisper.py

Co-authored-by: Sanchit Gandhi <93869735+sanchit-gandhi@users.noreply.github.com>

gau-nernst commented:
> Very nice @gau-nernst - especially the Flax init which is really clean now 👌 Could the PT init go in `_init_weights`? Otherwise it all looks good to me!

`_init_weights()` does not see the module name, it only sees the module itself. To make `_init_weights()` recognize encoder positional embedding, we probably need to set a private attribute to the module.

gau-nernst committed:
move sinusoidal init to _init_weights

gau-nernst commented:
I put sinusoidal init in `_init_weights()` but in a different way. Relying on the fact that `nn.Module.apply()` will traverse the children in a depth-first search manner (leaf modules will be applied first), if we check for Whisper encoder in `_init_weights()`, it will override the default initialization for positional embeddings.

```python
    def apply(self: T, fn: Callable[['Module'], None]) -> T:
        ...
        for module in self.children():
            module.apply(fn)
        fn(self)
        return self
```

sanchit-gandhi commented on src/transformers/models/whisper/modeling_whisper.py:
```diff
@@ -680,6 +680,10 @@ def _init_weights(self, module):
             module.weight.data.normal_(mean=0.0, std=std)
             if module.padding_idx is not None:
                 module.weight.data[module.padding_idx].zero_()
+        elif isinstance(module, WhisperEncoder):
```
1. I don't think there's a need for `torch.no_grad()` here - we should be able to run in the default torch context?
2. I'm not sure this necessarily works for initialising the modules - suppose that `WhisperEncoder` comes before the embedding layer in the list when we loop over all the modules: we first set the embeddings to sinusoids, but then set it to random again when we get to the embeddings

cc @ArthurZucker - do you know of a clean way to initialise a particular `nn.Embedding` layer with sinusoids, but not touch others?

gau-nernst replied:
1. It's a standard practice to use `torch.no_grad()` for weight init functions, since we don't want to capture computation graph of weight init. See https://github.com/pytorch/pytorch/blob/dac895c10ae0af86a5ca1183c1d08d0c00bc6c43/torch/nn/init.py#L17-L19 for example. However, in this case, since the sinusoidal embeddings have `.requires_grad = False`, it is possible to not use `torch.no_grad()` here. But I still prefer to have no grad for weight init.
2. As I stated above, `nn.Module.apply()` is depth-first search: the embedding layer (leaf module) will always come before `WhisperEncoder` (unless PyTorch changes this behaviour in the future, which is highly unlikely). 

I suppose to make it bullet-proof, the most straight-forward way is to set a private attribute. e.g.

```python
# in WhisperEncoder.__init__()
        self.embed_positions = nn.Embedding(self.max_source_positions, embed_dim)
        self.embed_positions.requires_grad_(False)
        self.embed_positions._hf_sinusoids_init = True

# in _init_weights()
        elif isinstance(module, nn.Embedding):
            if getattr(module, "_hf_sinusoids_init", False):
                # sinusoids init
            else:
                # normal embedding init
```

sanchit-gandhi replied:
Okay cool - thanks for the detailed explanations @gau-nernst! Good with me to keep as is then!

sanchit-gandhi reviewed (approved):
Thanks for iterating here @gau-nernst and for the fruitful comment discussions! Requesting a TF review from @Rocketknight1 and maintainer review from @ArthurZucker. Thanks both!

Rocketknight1 reviewed (approved):
TF code looks good to me! Doing it either this way or creating a `tf.constant` should both work.

ArthurZucker reviewed (approved):
Thanks for your contribution! 😉
